pkg/weather-client/client.go

```diff
@@ -0,0 +1,112 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+)
+
+const (
+	// used for retrieval in the noaa-client
+	weatherDotGovAPIEndpoint = "https://api.weather.gov"
+)
+
+type Elevation struct {
+	UnitCode string  `json:"unitCode"`
+	Value    float64 `json:"value"`
+}
+
+type Dewpoint struct {
+	UnitCode string  `json:"unitCode"`
+	Value    float64 `json:"value"`
+}
+
+type ProbabilityOfPrecipitation struct {
+	UnitCode string `json:"unitCode"`
+	Value    int    `json:"value"`
+}
+
+type RelativeHumidity struct {
+	UnitCode string `json:"unitCode"`
+	Value    int    `json:"value"`
+}
+
+type Period struct {
+	Dewpoint                   Dewpoint                   `json:"dewpoint"`
+	EndTime                    time.Time                  `json:"endTime"`
+	Icon                       string                     `json:"icon"`
+	IsDaytime                  bool                       `json:"isDaytime"`
+	Name                       string                     `json:"name"`
+	Number                     int                        `json:"number"`
+	ProbabilityOfPrecipitation ProbabilityOfPrecipitation `json:"probabilityOfPrecipitation"`
+	RelativeHumidity           RelativeHumidity           `json:"relativeHumidity"`
+	ShortForecast              string                     `json:"shortForecast"`
+	StartTime                  time.Time                  `json:"startTime"`
+	Temperature                int                        `json:"temperature"`
+	TemperatureTrend           interface{}                `json:"temperatureTrend"`
+	TemperatureUnit            string                     `json:"temperatureUnit"`
+	WindDirection              string                     `json:"windDirection"`
+	WindSpeed                  string                     `json:"windSpeed"`
+}
+
+type Properties struct {
+	Elevation         Elevation `json:"elevation"`
+	ForecastGenerator string    `json:"forecastGenerator"`
+	GeneratedAt       time.Time `json:"generatedAt"`
+	Periods           []Period  `json:"periods"`
+	Units             string    `json:"units"`
+	UpdateTime        time.Time `json:"updateTime"`
+	Updated           time.Time `json:"updated"`
+	ValidTimes        string    `json:"validTimes"`
+}
+
+type ForecastResponse struct {
+	Properties Properties `json:"properties"`
+	Type       string     `json:"type"`
+}
+
+type WeatherClient interface {
+	GetForecast(xCoordinate, yCoordinate float32) (*ForecastResponse, error)
+}
```

Comment:
got it, this is what you meant by testability. NoaaClient will implement the weather client and the consumer tests can just mock whatever weather client they want.

Reply:
exactly 👍.